dask/array/tests/test_gufunc.py

```diff
@@ -161,8 +161,8 @@ def foo(x, bar):
         assert bar == 2
         return x
 
-    ret = apply_gufunc(foo, "()->()", 1.0, output_dtypes="f", bar=2)
-    assert_eq(ret, np.array(1.0, dtype="f"))
+    ret = apply_gufunc(foo, "()->()", 1.0, output_dtypes=float, bar=2)
+    assert_eq(ret, np.array(1.0, dtype=float))
```

Comment:
This was failing because the array chunks were float64, while meta expected float32 (`"f"`). I'm not sure why it ever passed before?

Reply:
@pentschev I dug into this and found a number of concerning things:
* On `main`, despite saying `output_dtypes="f"` (float32), `ret.dtype` is float64! As I mentioned in the description, `output_types` is currently actually ignored unless inferring the dtype from `func` fails. Since calling the `foo` function with a float64 succeeds (the `1.0` input becomes NumPy float64), `apply_gufunc` returns a float64 array, even though you told it to do float32.
* **_Then_, `assert_eq` doesn't actually check dtypes?!** Turns out there is actually nothing in the `assert_eq` logic to check whether the dtypes of the two computed arrays match—just lots of logic to check that each one's meta matches its result. So the fact that `ret` and the expected `np.array(1.0, dtype="f")` have different dtypes is ignored, and the test passes. @jrbourbeau is this well-known? Should I open an issue about it? Seems like misleading behavior.

Since my changes now respect `output_dtypes`, the resulting array's `meta` (float32) does not match its computed dtype (float64, from the `1.0` input value), and `assert_eq` complained about that meta mismatch until I made the change now in this PR.

So basically, the old test was wrong: the output dtype in this case is in fact float64, so passing `output_dtype="f"` should have resulted in an error from `assert_eq` about a meta mismatch, except that `output_dtype` was ignored.

Reply:
> * On `main`, despite saying `output_dtypes="f"` (float32), `ret.dtype` is float64! As I mentioned in the description, `output_types` is currently actually ignored unless inferring the dtype from `func` fails. Since calling the `foo` function with a float64 succeeds (the `1.0` input becomes NumPy float64), `apply_gufunc` returns a float64 array, even though you told it to do float32.

To be honest, I'm not certain what's the exact behavior we want/users expect here. But if `output_dtypes` is specified, shouldn't `apply_gufunc` cast to the `output_dtypes`? Note that we do that in other places, such as [`partial_reduce`](https://github.com/dask/dask/blob/f547f081572cef0d2f99868d930d9256d2d9a00f/dask/array/reductions.py#L339). That would be, of course, only for the case where `meta` is computed automatically, but the user wants to enforce the `output_dtypes` nevertheless.

> * **_Then_, `assert_eq` doesn't actually check dtypes?!** Turns out there is actually nothing in the `assert_eq` logic to check whether the dtypes of the two computed arrays match—just lots of logic to check that each one's meta matches its result. So the fact that `ret` and the expected `np.array(1.0, dtype="f")` have different dtypes is ignored, and the test passes. @jrbourbeau is this well-known? Should I open an issue about it? Seems like misleading behavior.

Interesting catch, and I would say we should indeed open an issue. Even though I don't recall if there's a reason why we don't do it, it's worth at least discussing that, in case that is in fact a bug.

Reply:
I aggree with your assessment that `assert_eq` doesn't check computed dtype against meta dtype. That'd be a nice thing to add :)